### ISSUE TITLE
False positive overload errors after 1.1.226

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#3136]
After upgrading to 1.1.226, I'm noticing some strange issues with some overloaded functions/methods. Just saw this at my editor:

![Screenshot from 2022-03-02 15-41-03](https://user-images.githubusercontent.com/134025/156427771-d8d1ec14-d48f-464a-8a7b-102a4a1f948b.png)

The screenshot is from this line of this project: https://github.com/blb-ventures/strawberry-django-plus/blob/master/strawberry_django_plus/relay.py#L495 (I simplified the code to post the screenshot. but if you clone the project, do a `poetry install` in it to install the dependencies and run pyright, it should display the exact same issue)

The intention of the overload is that, if I pass `required=True`, it would raise an error if the return value would be `None` and thus, I want the typing to identify that it is not optional. Otherwise, passing either `required=False` or not passing anything (it defaults to `required=False`) would type the return value as optional.

obs. I could fix the error at t he screenshot by changing the second overload to `required: Literal[False] = ...`, but then I had this error at a place calling that method:

![Screenshot from 2022-03-02 15-54-16](https://user-images.githubusercontent.com/134025/156429062-86869818-731d-4469-98fc-9c1ac016e54a.png)

Also, I was trying to replicate the issue to post a simpler piece of code here, but not only I wasn't able to reproduce the issue but also I found another strange thing with it:

```python
from typing import Literal, Optional, overload

@overload
def get_something(required: Literal[True]) -> str:
    ...

@overload
def get_something(required: bool = ...) -> Optional[str]:
    ...

def get_something(required=False):
    if required:
        return ""
    return None
```

Running pyright on that gives me:

```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx.py
  /tmp/xxx.py:14:5 - error: Overloaded function implementation is not consistent with signature of overload 1
    Function return type "str" is incompatible with type "Literal[''] | None"
      Type "str" cannot be assigned to type "Literal[''] | None"
        Type cannot be assigned to type "None"
        "str" cannot be assigned to type "Literal['']" (reportGeneralTypeIssues)
  /tmp/xxx.py:14:5 - error: Overloaded function implementation is not consistent with signature of overload 2
    Function return type "str | None" is incompatible with type "Literal[''] | None"
      Type "str | None" cannot be assigned to type "Literal[''] | None"
        Type "str" cannot be assigned to type "Literal[''] | None"
          Type cannot be assigned to type "None"
          "str" cannot be assigned to type "Literal['']" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
Completed in 0.723sec
```

**Comment by erictraut** [comment]
Could you provide a minimal self-contained repro?

For the code sample that you posted, you need to provide a `-> Optional[str]` return type annotation to the implementation.

**Comment by erictraut** [comment]
I looked at the code in your project, and pyright is doing the correct thing in reporting an overlapping overload. Mypy emits the same error.

**Comment by bellini666** [comment]
Hey @erictraut , sorry but I'm failing to understand what is wrong in both situations. Maybe I'm missing something here...

You say that the first case the overlapping report is correct. But what would be the correct way of typing that then? Note in my explanation that I tried typing the second function with `Literal[False] = ...`, which solved the problem, but then I could not call that function anymore with a `bool` argument, only with literals `True` and `False`.

Also, note that in the code sample, unrelated to the other issue there, I have literally the same situation that is not being reported as overlapping. How is that different?

And for the code sample, even though adding a `-> Optional[str]` fixed the issue, I just think it is weird that the inferred literal string is not compatible with `str`. If it was the other way around, sure. But any literal string is `str` by extension.

**Comment by erictraut** [comment]
A type checker cannot always determine whether an argument is a literal or not. There will be times when someone passes `False` to the function but it is statically seen as a `bool`. As such, any overload that accepts a `bool` should return a supertype of an overload that accepts a `Literal[False]`. The [mypy documentation](https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-calls-to-overloads) may help you understand better.

**Comment by bellini666** [comment]
I get how it works, but that's not my point. Really sorry if I'm failing to communicate what I mean here, I was just trying to produce a counter argument.

Let me try to summarize my points in topics and in a clearer way:

1) At my code the overloads don't overlap. If the second function that is typed as `bool` came before `Literal[True]` then I would agree with you, but not in the way it is typed.

2) If you say that my code do indeed overlap, so should the 2 overloads of my code sample, but they are not being reported as such.

**Comment by erictraut** [comment]
Your overloads do overlap. If someone passes `Literal[True]`, the first overload will be used. If the user passes a `bool` value, which might be a `True`, the second overload will be used. The two overloads return different types, which means that if someone passes `True` but the static type checker cannot determine that it's `True`, it will evaluate a different return value than the first overload. 

If you want to provide more specific return types for an argument of `Literal[True]` and `Literal[False]`, that's fine, but the more general `bool` should return a superset of the types returned by the more specific overloads.

**Comment by erictraut** [comment]
I'm not exactly sure what you're trying to do here, but I suspect it's something like this:

```python
@overload
def foo(val: Literal[True]) -> float: ...

@overload
def foo(val: Literal[False]) -> str: ...

@overload
def foo(val: bool) -> float | str: ...
```

Note that the fallback overload (the third one) returns a superset of the types of the more specific overloads that come before it.

**Comment by bellini666** [comment]
Ok, an example might be better:

```python
from typing import Literal, Optional, cast, overload


def some_resource() -> Optional[str]:
    # This could be a request or anything that can be either str or None
    ...


@overload
def foo(required: Literal[True]) -> str:
    ...


@overload
def foo(required: bool = ...) -> Optional[str]:
    ...


def foo(required: bool = False) -> Optional[str]:
    retval = some_resource()
    if required is None:
        raise Exception("retval from some_resource expected, received None")

    return retval


reveal_type(foo(required=True))
reveal_type(foo(required=False))
reveal_type(foo(required=cast(bool, object())))
reveal_type(foo())
```
Running pyright gives me:
```
❯ pyright xxx.py 
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx.py
  /tmp/xxx.py:27:13 - information: Type of "foo(required=True)" is "str"
  /tmp/xxx.py:28:13 - information: Type of "foo(required=False)" is "str | None"
  /tmp/xxx.py:29:13 - information: Type of "foo(required=cast(bool, object()))" is "str | None"
  /tmp/xxx.py:30:13 - information: Type of "foo()" is "str | None"
0 errors, 0 warnings, 4 informations 
Completed in 0.591sec
```

This is what I was expecting from my code, and was also its behavior before 1.1.226. Note that the code covers the 4 possibilities:

1) `Literal[True]` falls into the first option and types the retval as `str`
2) `Literal[False]` is not `Literal[True]`, so as a bool it falls into the second option and types the retval as `str | None`
3) `bool` also falls into the second option
4) Not providing the argument falls into the second case, because it is the one that has `= ...` (the first one would have to be provided for sure)

This sample has literally the same typing logic as my code and is also producing the exact expected result I was expecting and also not giving me any overlapping warnings.

**Comment by erictraut** [comment]
No, this sample doesn't have the same typing logic as your code in the screenshot above. This sample is returning a supertype for the more general overload. `str | None` is a supertype of `str`.

In your screenshot above, the more specific overload is returning `Awaitable[NodeType]`, but the more general overload is returning `Awaitable[NodeType | None]`. The latter is not a supertype of the former. You would need to include `Awaitable[NodeType]` in the union returned by the more general case to eliminate the incompatible overlapping overload error.

**Comment by bellini666** [comment]
What about now:

```python
from typing import Any, Literal, Optional, cast, overload


def some_resource() -> Optional[str]:
    # This could be a request or anything that can be either str or None
    ...


@overload
def foo(required: Literal[True]) -> str:
    ...


@overload
def foo(required: bool = ...) -> float:
    ...


def foo(required: bool = False):
    return cast(Any, object())


reveal_type(foo(required=True))
reveal_type(foo(required=False))
reveal_type(foo(required=cast(bool, object())))
reveal_type(foo())
```

```
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx.py
  /tmp/xxx.py:23:13 - information: Type of "foo(required=True)" is "str"
  /tmp/xxx.py:24:13 - information: Type of "foo(required=False)" is "float"
  /tmp/xxx.py:25:13 - information: Type of "foo(required=cast(bool, object()))" is "float"
  /tmp/xxx.py:26:13 - information: Type of "foo()" is "float"
0 errors, 0 warnings, 4 informations 
Completed in 0.583sec
```

Still no issues and the more general overload `float` is not a superset of the specific `str`

**Comment by bellini666** [comment]
> In your screenshot above, the more specific overload is returning Awaitable[NodeType], but the more general overload is returning Awaitable[NodeType | None]. The latter is not a supertype of the former. You would need to include Awaitable[NodeType] in the union returned by the more general case to eliminate the incompatible overlapping overload error.

Just tried to make the general case of my code a superset of the specific. Still the same issue:

![Screenshot from 2022-03-02 23-35-07](https://user-images.githubusercontent.com/134025/156485340-a57a54e4-d8aa-426d-9451-158e45cd8578.png)

**Comment by erictraut** [comment]
> Still no issues and the more general overload float is not a superset of the specific str

That's because you have not enabled the "reportOverlappingOverload" diagnostic check. If you enable it, an error is reported here.

```
Overload 1 for "foo" overlaps overload 2 and returns an incompatible type
```

All of this is working the way it is designed to. The bug is in your code. I'd appreciate it if you'd spend more time diagnosing your own bugs before filing bugs against pyright.

**Comment by bellini666** [comment]
> That's because you have not enabled the "reportOverlappingOverload" diagnostic check. If you enable it, an error is reported here.

Ok, that's totally my fault, sorry about that. Still, [this comment](https://github.com/microsoft/pyright/issues/3136#issuecomment-1057606383) contains a "fix" as you suggested , making the return of the general case a superset of the specific. Still it is being flagged as incompatible.

> All of this is working the way it is designed to. The bug is in your code. I'd appreciate it if you'd spend more time diagnosing your own bugs before filing bugs against pyright.

Look, I'm really just trying to be helpful to the project here...

But note that you didn't even reply to my arguments, closed the issue with the "your code is wrong because it is" argument and even said mypy emitted the same error when it does not (I checked. It shows a lot of errors in other places, but not that one).

Also, It is working the way it is designed to work? Sure, you will for sure know better than me! But it wasn't until last version. That code started raising issues after I upgraded to 1.1.226 and I did my customary check on all my projects that I like to keep with 0 pyright warnings/errors. Because of the reasons I stated here, and after looking at some documentation and doing some tests it really seemed to me like a bug, that's the reason I came here to report it.
